crates/dojo-bindgen/src/plugins/typescript_v2/mod.rs

```diff
@@ -0,0 +1,594 @@
+            "
+// Type definition for `{}` enum
+export enum {} {{
+    {}
+}}
+",
+            token.type_path,
+            token.type_name(),
+            fields
+        )
+    }
+
+    // Formats a system into a C# method used by the contract class
+    // Handled tokens should be a list of all structs and enums used by the contract
+    // Such as a set of referenced tokens from a model
+    fn format_system(system: &Function, handled_tokens: &[Composite]) -> String {
+        let args = system
+            .inputs
+            .iter()
+            .map(|arg| {
+                format!(
+                    "{}: {}",
+                    arg.0,
+                    if TypeScriptV2Plugin::map_type(&arg.1.type_name()) == arg.1.type_name() {
+                        arg.1.type_name()
+                    } else {
+                        TypeScriptV2Plugin::map_type(&arg.1.type_name())
+                    }
+                )
+            })
+            .collect::<Vec<String>>()
+            .join(", ");
+
+        let calldata = system
+            .inputs
+            .iter()
+            .map(|arg| {
+                let token = &arg.1;
+                let type_name = &arg.0;
+
+                match handled_tokens.iter().find(|t| t.type_name() == token.type_name()) {
+                    Some(t) => {
+                        // Need to flatten the struct members.
+                        match t.r#type {
+                            CompositeType::Struct => t
+                                .inners
+                                .iter()
+                                .map(|field| format!("props.{}.{}", type_name, field.name))
+                                .collect::<Vec<String>>()
+                                .join(",\n                    "),
+                            _ => type_name.to_string(),
+                        }
+                    }
+                    None => type_name.to_string(),
+                }
+            })
+            .collect::<Vec<String>>()
+            .join(",\n                ");
+
+        format!(
+            "async {pretty_system_name}({args}): Promise<void> {{
+        try {{
+            await this.execute(\"{system_name}\", [{calldata}])
+        }} catch (error) {{
+            console.error(\"Error executing {pretty_system_name}:\", error);
+            throw error;
+        }}
+    }}",
+            pretty_system_name = system.name.to_case(convert_case::Case::Camel),
+            // formatted args to use our mapped types
+            args = args,
+            system_name = system.name,
+            // calldata for execute
+            calldata = calldata
+        )
+    }
+
+    // Formats a contract file path into a pretty contract name
+    // eg. dojo_examples::actions::actions.json -> Actions
+    fn formatted_contract_name(contract_file_name: &str) -> String {
+        let contract_name =
+            contract_file_name.split("::").last().unwrap().trim_end_matches(".json");
```

Comment:
Is there a chance of conflict if the project has several contracts with the same name but not the same fully qualified path?

Reply:
Here's how I understood it, but might be wrong here, so open to suggestions!

This is the fully qualified name of a system `dojo_starter::systems::actions::actions`. I was under the impression that the last two elements (so `actions`, twice) were always exactly the same. And seeing as `dojo_starter` and `systems` are constant (for a project), I thought that there's no way the parsed contract name here can cause clashes.

But like I said, might be me misunderstanding!

Reply:
The latest `actions` is the module name (inside the Cairo file, the `dojo::contract` is on the top of `mod actions`.)
And then the other `actions` is the file (which is also a Cairo module). So we ends up with duplication here. So usually for contracts, we don't have 2 contracts in the same file.

For models, it is much more probable to have several models into one file.
So we could keep it like this for now, and we will see in a next round how we can ensure we don't have conflicts. Because `dojo_examples::module1::actions` will conflict with `dojo_examples::modules2::actions`. This is also part of the test revamp where we will add much more examples/cases to ensure we cover them correctly.
Also, we will add a namespace feature in the future, and using this, we will always ensure that only one contract can have a specific name into the namespace. Hence, the SDK will use this namespace to access the contract object.
We should also sync about this next week, as it's not even present in the dojo-core yet, but planned to be.